### PR TITLE
Fix: FCM Make and Rose Prune

### PR DESCRIPTION
**Pull request by wxtim**
Partial Response to [Metomi Rose issue 2445](https://github.com/metomi/rose/issues/2445):
[Metomi Rose Companion PR2461](https://github.com/metomi/rose/pull/2461) will be required to use the new functionality in this PR -

# Synopsis
Add a way of extracting Platform config for a task in a workflow from:
1. A Workflow definition: Given the names of the flow and the task.
2. A Workflow database: Given the names of the flow, the task and the cycle point.

# Summary of Work
- [x] Created a function `get_platform_from_task_def`.
  - [x] Added a basic set of tests.
  - [x] Added a test which checks that tasks with inherited features are OK.
  - [x] Added a test which checks that tasks with Cylc 7 style platform definitions work OK. 
- [x] Created a function `get_platform_from_task_job`.
  - [x] Added a basic set of tests.
  - [x] Added tests checking that the last submission of each job is selected

### CODE REVIEW

File: tests/test_platform_utils.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            return glbl_cfg
+
+        monkeypatch.setattr(pypath, _inner)
+
+    yield _mock
+    rmtree(tmp_path)
+
+
+@pytest.fixture(scope='session')
+def fake_flow():
```
Comment (MetRonnie) [inline tests/test_platform_utils.py:101]: Docstring (and possibly type annotation to show what it returns)?

File: tests/test_platform_utils.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    2. Set up a fake flow database in ``.service/db``.
+
+    Returns:
+        flow_name: Name of fake workflow.
+        flow_path: Path at which fake workflow is installed.
+    """
+    # Set up a config
+    flow_name: str = f'cylc-rose-platform-utils-test-{str(uuid4())[:6]}'
+    flow_path: Path
+    flow_path = Path(os.path.expandvars('$HOME/cylc-run')) / flow_name
```
Comment (MetRonnie) [inline tests/test_platform_utils.py:113]: (No need for a type annotation here, as mypy can infer the type for a simple case like this)

File: tests/test_platform_utils.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        b"cycle TEXT, name TEXT, submit_num INTEGER, platform_name TEXT);\n"
+        b"INSERT INTO task_jobs (cycle, name, submit_num, platform_name)"
+        b"    VALUES ('1', 'bar', 1, 'localhost');\n"
+        b"INSERT INTO task_jobs (cycle, name, submit_num, platform_name)"
+        b"    VALUES ('1', 'baz', 1, 'dairy');\n"
+        b"INSERT INTO task_jobs (cycle, name, submit_num, platform_name)"
+        b"    VALUES ('1', 'bar', 2, 'dairy');\n"
+        b"INSERT INTO task_jobs (cycle, name, submit_num, platform_name)"
+        b"    VALUES ('2', 'baz', 1, 'milk');\n"
+    )
```
Comment (MetRonnie) [inline tests/test_platform_utils.py:149]: Still need to remove the `tests/test_platform_utils/db` file

**Review by MetRonnie — APPROVED** [review]
👍

File: cylc/rose/platform_utils.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+# along with Rose. If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+"""Interfaces for Cylc Platforms for use by rose apps.
+"""
+from pathlib import Path
+from typing import Optional, Dict, Any
+
+from cylc.flow.config import SuiteConfig as FlowConfig   # Futureproof
+from cylc.flow.rundb import CylcSuiteDAO as CylcFlowDAO
+from cylc.flow.suite_files import parse_suite_arg as parse_flow_arg
```
Comment (datamel) [inline cylc/rose/platform_utils.py:25]:
```suggestion
from cylc.flow.config import WorkflowConfig as FlowConfig   # Futureproof
from cylc.flow.rundb import CylcWorkflowDAO as CylcFlowDAO
from cylc.flow.workflow_files import parse_workflow_arg as parse_flow_arg
```

**Review by datamel — APPROVED** [review]
I have checked this branch out with sibling pr. Everything is working smoothly for me (I rebased @MetRonnie's suite-to-workflow branch and changed the imports as suggested in order to get it working).

**Comment by datamel** [comment]
Just a note to hold fire on merging this PR for now. https://github.com/metomi/rose/pull/2461 has problems with `rosie checkout`, confirmed by @wxtim.  Looking into this now.

**Review by MetRonnie — CHANGES_REQUESTED** [review]
Will need suite -> workflow changes

**Comment by MetRonnie** [comment]
Just need to squash the `squash!` commits. Btw, looks like something is up with the branch history, I'm seeing a range of commits twice

**Comment by wxtim** [comment]
> Just need to squash the `squash!` commits. Btw, looks like something is up with the branch history, I'm seeing a range of commits twice

Horrible merge. I think I've sauteed this now.

**Review by MetRonnie — APPROVED** [review]

**Comment by MetRonnie** [comment]
I see this is still open, does it need a re-review from @datamel ?

File: cylc/rose/platform_utils.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        flow: The name of the Cylc flow to be queried.
+        task: The name of the task to be queried.
+
+    Returns:
+        Platform Dictionary.
+    """
+    flow_name, flow_file = parse_workflow_arg(None, flow)
+    config = WorkflowConfig(flow, flow_file, None)
+    # Get entire task spec to allow Cylc 7 platform from host guessing.
+    task_spec = config.pcfg.get(['runtime', task])
```
Comment (oliver-sanders) [inline cylc/rose/platform_utils.py:47]: I can't remember what `pcfg` is supposed to be, does this one work with inheritance?

Comment (wxtim) [reply cylc/rose/platform_utils.py:47]: No idea, I'll have to have a play with it...

Comment (oliver-sanders) [reply cylc/rose/platform_utils.py:47]: Ta, just worth running through a quick inheritance example to make sure.

**Review by datamel — COMMENTED** [review]
When checked out this with its sibling, the tests now pass for me. This is limited by some skipping rose tests which skipped due to `"[t]job-host" not defined or not available` - I am unsure why, since I believe my config looks ok but other reviewers may want to get them to run without skipping. Given the tests what are skipping are some app-prune and fcm-make ones, I'd feel much happier if I were able to get these to run.

**Comment by MetRonnie** [comment]
I get `skipped: "[t]job-host" not defined` for `t/rose-app-run/12-file-incr-3.t`

**Review by datamel — COMMENTED** [review]
Hold fire on merging this one. Tests failing for me, @wxtim is aware and will work on the fix.

**Comment by datamel** [comment]
> I get `skipped: "[t]job-host" not defined` for `t/rose-app-run/12-file-incr-3.t`. (Have just tried `t/rose-app-run/*` so far)

My `rose.conf` now has these running ok, with @oliver-sanders fix from teams, so, fingers crossed, you should be able to copy that and get the tests to run.

**Comment by MetRonnie** [comment]
> > I get `skipped: "[t]job-host" not defined` for `t/rose-app-run/12-file-incr-3.t`. (Have just tried `t/rose-app-run/*` so far)
> 
> My `rose.conf` now has these running ok, with @oliver-sanders fix from teams, so, fingers crossed, you should be able to copy that and get the tests to run.

Working now

File: cylc/rose/platform_utils.py (hunk 5 above)
Comment (wxtim) [reply cylc/rose/platform_utils.py:47]: Looks OK.

**Review by datamel — APPROVED** [review]
All tests now passing. Merging both PRs.

**Pull request merged**